packages/kas/src/__tests__/compilation_test.js

```diff
@@ -0,0 +1,100 @@
+import _ from "underscore";
+
+import * as KAS from "../index.js";
+
+expect.extend({
+    toCompileAs(
+        input: string,
+        expected: number,
+        vars?: {[string]: string | number} = {},
+        functions?: $ReadOnlyArray<string>,
+    ) {
+        const func = KAS.parse(input, {functions: functions}).expr.compile();
+
+        const actual = func(vars);
+
+        return Math.abs(actual - expected) < 1e-9
+            ? {pass: true}
+            : {
+                  pass: false,
+                  message: () =>
+                      `${input} should evaluate to ${expected}, was ${actual}; by func: ${func.toString()}`,
+              };
+    },
+});
+
+describe("compilation", () => {
+    test("empty", () => {
+        expect("").toCompileAs(0);
+    });
+
+    test("simple expressions", () => {
+        expect("1+2+3+4").toCompileAs(10);
+        expect("1+2-3+4").toCompileAs(4);
+        expect("1*2*3*4").toCompileAs(24);
+        expect("1*2/3*4").toCompileAs(2 + 2 / 3);
+        expect("4^3^2^1").toCompileAs(262144);
+        expect("-1").toCompileAs(-1);
+        expect("--1").toCompileAs(1);
+        expect("---1").toCompileAs(-1);
+        expect("2^-2").toCompileAs(0.25);
+        expect("8^(1/3)").toCompileAs(2);
+        expect("0^0").toCompileAs(1);
+        expect(".25*4").toCompileAs(1);
+        expect("ln e").toCompileAs(1);
+        expect("log 10").toCompileAs(1);
+        expect("log_2 2").toCompileAs(1);
+    });
+
+    test("variable expressions", () => {
+        expect("x").toCompileAs(3, {x: 3});
+        expect("x^2").toCompileAs(9, {x: 3});
+        expect("(x^2+y^2)^.5").toCompileAs(5, {x: 3, y: 4});
+        expect("log x_0").toCompileAs(1, {x_0: 10});
+        expect("log x_0 + log x_1").toCompileAs(3, {x_0: 10, x_1: 100});
+        expect("log x_42").toCompileAs(1, {x_42: 10});
+        expect("x_a + x_bc").toCompileAs(7, {x_a: 1, x_b: 2, c: 3});
+    });
+
+    test("function expressions", () => {
+        expect("f(2)").toCompileAs(
+            4,
+            {
+                f: function (x) {
+                    return 2 * x;
+                },
+            },
+            ["f"],
```

Comment:
I wonder if the assertion could eliminate the need for the last array of function names and just deduce that based on the vars being passed in: ie: `const fns = Object.keys(vars).filter(k => typeof vars[k] === 'function')` Maybe?